proposals/wg-image-compatibility.md

```diff
@@ -0,0 +1,104 @@
+These are usually described in the release notes or installation guides.
+We believe we have identified the missing puzzle in the OCI standard, the image compatibility specification, which is a key feature that describes the requirements of special containerized applications that have hard requirements for specific kernel versions, configurations, modules, or out-of-tree drivers.
+
+For example, a container application free5GC UPF requiring the gtp5g kernel module will only be compatible with kernel 5.0.0-23-generic or 5.4.x due to the module's hard kernel version requirements - https://free5gc.org/guide/3-install-free5gc/.
+If SR-IOV or GPU is required by the container, other requirements have to be specified against the host OS.
+Kata Containers use cases are the best example: https://github.com/kata-containers/kata-containers/tree/main/docs/use-cases.
+
+
+The described incompatibility problem affects bare-metal (e.g. high performance computing) and distributed systems use cases.
+
+In multi-cloud, hybrid cloud, or even private cloud scenarios, one containerized cloud-native application will need to run on different Kubernetes distributions that use different host operating systems on different worker nodes.
+
+For bare-metal and cloud operating systems can use different kernel versions with different configurations.
+Different operating system distributions may integrate different devices and their corresponding drivers, even for the same device different versions of drivers may be used.
+Therefore, different operating system distributions are not identical.
+Even within the same OS distro, different releases are also not identical.
+
+An image compatibility specification would help container image authors describe compatibility requirements in a standard way.
+The specification will be uploaded with the image to the image registry.
+This makes hard container compatibility requirements discoverable, programmable, and will support different consumers and cover use cases where the application requires a specific compatible environment.
+
+For further reference please follow:
+* [OCI Image Compatibility - the spec reasoning](https://docs.google.com/presentation/d/1F9GnCm2sULuyTJ5BEFZlL8Qjab81DK7g9Oy6qBfe5Qs/edit?usp=sharing)
+* [Revisit container and host OS compatibility](https://docs.google.com/document/d/1lzwh8DGMu5vXXHwJmnewYIMffkcOEvH8owX4UYjRcw0/edit#heading=h.fcxt9vheg92c)
+
+## Scope
+
+* Define image compatibility spec that describes special host OS requirements of containerized application.
+* The spec should describe container requirements for Linux, illumos, FreeBSD and Windows.
+* The final list of supported fields (like kernel modules, configuration, out-of-tree drivers etc.) is to be determined by the working group for each supported OS.
+* Make the change as minimally invasive as possible to the image, runtime and distribution specs - that could be achieved over delivering compatibility artifact type.
```

Comment:
I think I might be missing something -- how are the goals of an "image compatibility" spec different from image-spec, especially re: the `platform` object described [here](https://github.com/opencontainers/image-spec/blob/main/image-index.md)?

My first thought would be that the output of this WG would be to standardize some values and/or new fields there, not produce a new artifact type.

Reply:
my personal opinion (although I haven't spend time with the stakeholders/proposers yet) is that this WG should flesh out the use cases and determine/propose what is the best model to carry this kind of informational alongside/with an image definition. Given I was involved in the original Docker v2 discussions about the platform object when multiplatform images were first evolving, the `features` flag was meant to potentially hold the kind of descriptive information that runtimes could use to determine some of the compatibility issues noted in this WG proposal. However, there has never (to my knowledge) been any attempt to define how to standardize the set of disparate information (CPU flags/features? GPU available? certain kernel modules available?) and encode it in that object entity. I would hope this WG can discuss the tradeoffs of fields available versus new fields versus a new artifact before pre-determining the specific solution that is necessary to meet all the use cases/requirements.

Reply:
For a quick summary of my thoughts - compatibility is much more than platform, and actually the data can be enormous. The reason to have a separate spec would be to 1. account for that, 2. allow for using it to be optional (without putting burden on tools to support it and users to download data they don't need), and 3. to allow for (possibly) extension to other aspects of image compatibility, such as on the level of the library. When you especially consider that third point, and realize ABI compatibility goes into binary analysis of storing not just high level metadata but also (potentially) symbols / names and types and metadata about compilers, hardware, and the kernel, then it becomes a much harder (and larger) problem to do the assessment. I'm not saying that will be the ultimate outcome here, but I it's such an important issue that I don't think we want to close the door that it couldn't be an idea supported (e.g., if someone made a custom compatibility spec for it). I also don't think we want to, off the bat, mix up the purposes of these two (imho) very different things. The platform specific tags are used to determine, on a high level, which to pull, but not to assess compatibility of those detailed things.

Reply:
> I think I might be missing something -- how are the goals of an "image compatibility" spec different from image-spec, especially re: the `platform` object described [here](https://github.com/opencontainers/image-spec/blob/main/image-index.md)?
> 
> My first thought would be that the output of this WG would be to standardize some values and/or new fields there, not produce a new artifact type.

One of the arguments against extending the image spec for compatibility was to not make an image container release lifecycle dependend on third-party software versions in all cases. For example, whenever a new cuda driver release happens, the image author must test the driver and create a new image build to update the compatibility spec. This is not optimal unless user application changes are required to work fine with the driver. Users would be forced to release a new version just to update the metadata.

We thought to use artifact to have a distinct entity that has self-contained structure and is easily interchangeable to describe image compatibilty.

Additionally, as we explored the use cases, we realized that the compatibility spec can be a slightly more complex structure.
We will need to provide users to create some sort of abstraction to cover different vendors (different configuration) that enables similar features.

> I would hope this WG can discuss the tradeoffs of fields available versus new fields versus a new artifact before pre-determining the specific solution that is necessary to meet all the use cases/requirements.

That's fair. Let me adjust the scope to reflect this. Looks like me, @ChaoyiHuang and @vsoch are in the artifact camp, but let's have a discussion within the wg to hear others.

Reply:
> I would hope this WG can discuss the tradeoffs of fields available versus new fields versus a new artifact before pre-determining the specific solution that is necessary to meet all the use cases/requirements.

totally +1

Reply:
@estesp I removed any suggestions for potential implementation from the document and focused only on the goals.